### PR TITLE
[HWArith] Propagate (and improve) sv.namehints in HW lowering

### LINKED ISSUE #4225: [HWArith] Propagate `sv.namehint` during lowering
Ideally, all of the operations created in the lowering to comb would have the original op's `sv.namehint` or some derivation thereof. This would __drastically__ help with debugging.

### PR DESCRIPTION
**Pull request by mortbopet**
This commit propagates attributes from source ops to the "main" generated comb operations. Furthermore, tries to create meaningful names for bit extractions and width extensions.

Fixes #4225.

### CODE REVIEW

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp [hunk 1]
```diff
@@ -222,8 +259,11 @@ struct ICmpOpLowering : public OpConversionPattern<ICmpOp> {
     Value rhsValue = extendTypeWidth(rewriter, loc, adaptor.getRhs(), cmpWidth,
                                      rhsType.isSigned());
 
-    rewriter.replaceOpWithNewOp<comb::ICmpOp>(op, combPred, lhsValue, rhsValue,
-                                              false);
+    auto newOp = rewriter.replaceOpWithNewOp<comb::ICmpOp>(
+        op, combPred, lhsValue, rhsValue, false);
+    auto predAttr = newOp.getPredicateAttr();
+    newOp->setAttrs(op->getAttrs());
+    newOp->setAttr(newOp.getPredicateAttrName(), predAttr);
```
Comment (7FM) [inline lib/Conversion/HWArithToHW/HWArithToHW.cpp:266]: Not sure if I am a fan of this approach. If the operations get new internal attributes (or get renamed), these would be propagated too, wouldn't it?
Is there a particular reason why you didn't choose an explicit attribute propagation as in: https://github.com/llvm/circt/blob/b5e6fa769a580faa1c5ebda29d8a01c45633bc08/lib/Dialect/Comb/CombFolds.cpp#L61-L64

Comment (teqdruid) [reply lib/Conversion/HWArithToHW/HWArithToHW.cpp:266]: I agree. It's reasonable (and relatively easy) to just copy the dialect attributes. See `op.(get|set)DialectAttrs`.

Comment (teqdruid) [reply lib/Conversion/HWArithToHW/HWArithToHW.cpp:266]: It would also avoid having to re-set the pred attr.

**Review by teqdruid — APPROVED** [review]
Plz switch to copying the dialect attributes only. Then I think you're good to go.

File: lib/Conversion/HWArithToHW/HWArithToHW.cpp (hunk 1 above)
Comment (mortbopet) [reply lib/Conversion/HWArithToHW/HWArithToHW.cpp:266]: Agree on all points - will change to get/setDialectAttrs.

**Pull request merged**